### PR TITLE
add docs/admin/cluster-components.md

### PR DESCRIPTION
**Pull request by lavalamp**

### CODE REVIEW

**Comment by lavalamp** [comment]
@davidopp @krousey @bgrant0607

File: docs/admin/cluster-components.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+# Kubernetes Cluster Admin Guide: Cluster Components
+
+This document outlines the various binary components that need to run to
+deliver a functioning Kubernetes cluster.
+
+## Master Components
+
+Master components are those that provide the cluster's control plane, make
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: Making global decisions and detecting and responding to cluster events are part of the cluster's control plane, so this sentence isn't really parallel. I'd write something like "Master components are those that provide the cluster's control plane. For example, master components are responsible for making global decisions about the cluster (e.g., scheduling), and detecting and responding to cluster events (e.g., starting up a new pod when a replication controller's 'replicas' field is unsatisfied).

File: docs/admin/cluster-components.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+multi-master-VM setup.)
+
+Even in the future, when Kubernetes is fully self-hosting, it will probably be
+wise to only allow master components to schedule on a subset of nodes, to limit
+co-running with user-run pods, reducing the possible scope of a
+node-compromising security exploit.
+
+### kube-apiserver
+
+```kube-apiserver``` is the global point of contact for the api. It is designed to
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: I'm not sure everyone will understand "global point of contact for the api." Maybe "exposes the Kubernetes API; it is the  front-end for the Kubernetes control plane" (feel free to toss "global" into that if you want; that's not the part I was trying to change)

File: docs/admin/cluster-components.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+scale horizontally (i.e., one scales it by running more of them).
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: Point people to the HA doc again for more details?

File: docs/admin/cluster-components.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+### etcd
+
+```etcd``` is used as Kubernetes' backing store. All cluster data is stored here.
+Proper administration of a Kubernetes cluster includes a backup plan for etcd's
+data.
+
+### kube-controller-manager
+
+```kube-controller-manager``` is a binary that runs controllers, or the background
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: or -> which are

File: docs/admin/cluster-components.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+processes that handle routine tasks in the cluster.
+
+These controllers include:
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: I would suggest indicating how many there typically are of each, at least distinguish which only have one instance (e.g. node controller) vs. may have many instances (e.g. replication controller). For the latter category, explain how many there are (e.g. for replication controller I would rewrite as "For each ReplicationController object in the system, there is a replication controller that is responsible for maintaining the correct number of pods for a corresponding ReplicationController.")

File: docs/admin/cluster-components.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+* Node Controller
+ * Responsible for noticing & responding when nodes go down.
+* Replication Controller
+ * Responsible for maintaining the correct number of pods for every replication
+   controller in the system.
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: controller object (it's a bit circular to say the replication controller is responsible for maintaining the correct number of pods for every replication controller)

File: docs/admin/cluster-components.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+### etcd
+
+```etcd``` is used as Kubernetes' backing store. All cluster data is stored here.
+Proper administration of a Kubernetes cluster includes a backup plan for etcd's
+data.
+
+### kube-controller-manager
+
+```kube-controller-manager``` is a binary that runs controllers, or the background
+processes that handle routine tasks in the cluster.
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: can you make it clearer that the controllers run as part of the kube-controller-manager process, not as separate processes? Maybe something like "is a binary that runs controllers, which are the background tasks that handle routine operations in the cluster. The controllers run as threads in the kube-controller-manage rprocess."

File: docs/admin/cluster-components.md [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+   controller in the system.
+* Endpoints Controller
+ * Populates the Endpoints object (i.e., join Services & Pods).
+* Service Account & Token Controllers
+ * Create default accounts and API access tokens for new namespaces.
+* ... and others.
+
+### kube-scheduler
+
+```kube-scheduler``` watches newly created pods that have no node, and selects a node
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: no node -> no node assigned

File: docs/admin/cluster-components.md [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+for them to run on.
+
+## Node components
+
+Node components run on every node, maintaining running pods and providing them
+the Kubernetes running environment.
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: running -> runtime

File: docs/admin/cluster-components.md [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+```docker``` is of course used for actually running containers.
+
+### rkt
+
+```rkt``` is supported experimentally as an alternative to docker.
+
+### monit
+
+```monit``` is a lightweight montoring system for keeping kubelet and docker
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: monitoring -> babysitting?

Comment (davidopp) [inline docs/admin/cluster-components.md]: or "process babysitting"

File: docs/admin/cluster-components.md [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+## Node components
+
+Node components run on every node, maintaining running pods and providing them
+the Kubernetes running environment.
+
+### kubelet
+
+```kubelet``` is the primary node agent. It watches for pods that have been
+assigned to its node and runs them.
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: runs them -> tells docker or rkt to run them
Also mention the health checking functionality?

File: docs/admin/cluster-components.md [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+
+### kube-proxy
+
+```kube-proxy``` is responsible for maintaining network rules, enabling the
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: I'd reword this slightly, maybe something like "enables the Kubernetes service abstraction by maintaining network rules on the host and performing connection forwarding"

**Comment by davidopp** [comment]
I was going to suggest to link to docs/{kube-apiserver,kubelet-kube-scheduler,kube-controller-manager,kube-proxy} from your new section, but it seems those files aren't referred to from anywhere. If so, I'd suggest to just delete those files. But if you want to keep them, I'd suggest moving them into docs/admin and linking to them from your new section.

File: docs/admin/cluster-components.md (hunk 5 above)
Comment (lavalamp) [inline docs/admin/cluster-components.md]: >I would suggest indicating how many there typically are of each

No controller has more than one, though?

>For each ReplicationController object in the system, there is a replication controller that is responsible for maintaining the correct number of pods for a corresponding ReplicationController."

That's just not true, though; there's a single replication controller manager (replication controller controller?). We don't even start a thread per RC object. RC object is very unfortunately named for talking about the thing that actually implements it.

File: docs/admin/cluster-components.md (hunk 7 above)
Comment (lavalamp) [inline docs/admin/cluster-components.md]: expanded

File: docs/admin/cluster-components.md (hunk 5 above)
Comment (davidopp) [inline docs/admin/cluster-components.md]: Yeah, never mind, I was confusing controllers and objects that are controlled.

File: docs/admin/cluster-components.md (hunk 6 above)
Comment (davidopp) [inline docs/admin/cluster-components.md]: Never mind this comment; your earlier clarification explained why you said this. Our terminology is pretty confusing though.

Comment (davidopp) [inline docs/admin/cluster-components.md]: Actually, I take that back. I think "replication controller object" makes it clearer

File: docs/admin/cluster-components.md (hunk 11 above)
Comment (lavalamp) [inline docs/admin/cluster-components.md]: expanded

**Comment by lavalamp** [comment]
Feedback mostly addressed; I'll look at moving/linking to those docs next.

**Comment by lavalamp** [comment]
I've moved those docs; I think they can be improved but maybe I'll do that in another PR. @davidopp let me know when you've reviewed my changes and I'll squash.

**Comment by davidopp** [comment]
Looking now.

File: docs/admin/cluster-components.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+
+Master components are those that provide the cluster's control plane. For
+example, master components are responsible for making global decisions about the
+cluster (e.g., scheduling), and detecting and responding to cluster events
+(e.g., starting up a new pod when a replication controller's 'replicas' field is
+unsatisfied).
+
+Master components could in theory be run on any node in the cluster. However,
+for simplicity, current set up scripts typically start all master components on
+the same VM. (But see [availability.md](availability.md) for an example
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: "and no user containers are allowed to run on this VM" ?

File: docs/admin/cluster-components.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+Even in the future, when Kubernetes is fully self-hosting, it will probably be
+wise to only allow master components to schedule on a subset of nodes, to limit
+co-running with user-run pods, reducing the possible scope of a
+node-compromising security exploit.
+
+### kube-apiserver
+
+[kube-apiserver](kube-apiserver.md) exposes the Kubernetes API; it is the front-end for the
+Kubernetes control plane It is designed to scale horizontally (i.e., one scales
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: missing period before It

File: docs/admin/cluster-components.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+```etcd``` is used as Kubernetes' backing store. All cluster data is stored here.
+Proper administration of a Kubernetes cluster includes a backup plan for etcd's
+data.
+
+### kube-controller-manager
+
+[kube-controller-manager](kube-controller-mangager.md) is a binary that runs controllers, which are the
+background threads that handle routine tasks in the cluster. Logically, each
+controller is a separate process, but to reduce the number of moving pieces in
+the system, they are all compiled into a single binary.
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: "single binary and run in a single process"

(in theory you could compile them into a single binary that forks off suprocesses or something -- just want to make it crystal clear)

File: docs/admin/cluster-components.md [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+
+### etcd
+
+```etcd``` is used as Kubernetes' backing store. All cluster data is stored here.
+Proper administration of a Kubernetes cluster includes a backup plan for etcd's
+data.
+
+### kube-controller-manager
+
+[kube-controller-manager](kube-controller-mangager.md) is a binary that runs controllers, which are the
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: misspelled link

**Comment by davidopp** [comment]
looks good except for the minor comments I just made

**Comment by lavalamp** [comment]
Comments addressed, and I added a section about addons which could use a look.

File: docs/admin/cluster-components.md [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+* ... and others.
+
+### kube-scheduler
+
+[kube-scheduler](kube-scheduler.md) watches newly created pods that have no node assigned, and
+selects a node for them to run on.
+
+### addons
+
+Addons are pods and services that implement cluster features. They aren't
```
Comment (davidopp) [inline docs/admin/cluster-components.md]: I think this paragraph could be rewritten a bit more concisely: "Addons are pods and services that implement cluster features. They don't run on the master VM, but currently the default setup scripts (link stuff) that make the API calls to create these pods and services do run on the master VM."

(the stuff about "run as pods within the cluster, not directly on VMs of the cluster" is confusing and I'd just drop it)

BTW should you say something about the kube-system namespace?

File: docs/admin/cluster-components.md [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+
+![WARNING](http://kubernetes.io/img/warning.png)
+![WARNING](http://kubernetes.io/img/warning.png)
+![WARNING](http://kubernetes.io/img/warning.png)
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+# Kubernetes Cluster Admin Guide: Cluster Components
+
```
Comment (bprashanth) [inline docs/admin/cluster-components.md]: mention static pods somewhere?

**Comment by lavalamp** [comment]
OK, last round of comments addressed, this should be ready to go.

File: docs/admin/cluster-components.md (hunk 18 above)
Comment (lavalamp) [inline docs/admin/cluster-components.md]: done

File: docs/admin/cluster-components.md (hunk 17 above)
Comment (lavalamp) [inline docs/admin/cluster-components.md]: done

**Comment by davidopp** [comment]
LGTM

File: docs/admin/cluster-components.md (hunk 18 above)
Comment (bprashanth) [inline docs/admin/cluster-components.md]: I meant more like daemons started one per node via manifests passed to the kubelet via --config, to clarify why docker ps shows a fluentd pod when they didn't start one, and deleting them will lead to the kubelet starting another

**Comment by lavalamp** [comment]
rebased

**Comment by wojtek-t** [comment]
It seems that not all tests are passing here:

"The command "PATH=$GOPATH/bin:./third_party/etcd:$PATH ./hack/verify-gendocs.sh" exited with 1."

**Comment by davidopp** [comment]
shippable says you need to run hack/run-gendocs.sh

**Comment by lavalamp** [comment]
rebased

File: docs/admin/cluster-components.md (hunk 18 above)
Comment (lavalamp) [inline docs/admin/cluster-components.md]: I did reference that it looked for static config files. Feel free to send a PR expanding on that if you still want more detail :)

**Pull request merged**